File: multiples_of_3_and_5.rb
# frozen_string_literal: true

#  Project Euler exercises - Problem 1
#   If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these
#  multiples is 23.
# Find the sum of all the multiples of 3 or 5 below 1000.
def solution
  array = (1..999).to_a
  new_array = Array.new
  solution = 0
   array.each do |i|
    if i%3 == 0 or i%5 == 0
      new_array.push(i)
    end
   end
  new_array.each { |j| solution+=j}
  puts(solution)
end

solution

File: rust/multiples_of_3_and_5.rs
// Problem:
// If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9.
// The sum of these multiples is 23. Find the sum of all the multiples of 3 or 5 below 1000.
fn main() {
    let f = &[3, 5];
    println!("{}", sum_of_multiples(1000, f));
}

pub fn sum_of_multiples(limit: u32, factors: &[u32]) -> u32 {
    (1..limit).filter(|&i| factors.iter().any(|&j| i % j == 0)).sum()
}